### PR TITLE
generic bf16 support

### PR DESCRIPTION
**Pull request by Qazalin**
the immediate problem this will solve is that Metal doesn't support PtrDTypes on bf16 https://github.com/tinygrad/tinygrad/actions/runs/7240422493/job/19723393442?pr=2794

### CODE REVIEW

File: test/test_dtype.py [hunk 1]
```diff
@@ -345,6 +346,29 @@ def test_broadcast_int(self):
     assert (Tensor.rand(4, 4, dtype=dtypes.float32) + 2).dtype == dtypes.float32
     assert (Tensor.rand(4, 4, dtype=dtypes.float64) + 2).dtype == dtypes.float64
 
+# for GPU, cl_khr_fp16 isn't supported
+# for LLVM, it segfaults because it can't link to the casting function
+# CUDACPU architecture is sm_35 but we need at least sm_70 to run fp16 ALUs
+@unittest.skipIf(Device.DEFAULT in ["GPU", "LLVM", "CUDA"] and CI, "fp16 broken in some backends")
+@unittest.skipIf(Device.DEFAULT == "TORCH", "torch doesn't have uint32")
+@unittest.skipIf(Device.DEFAULT == "WEBGPU", "webgpu's fp16 and uint16 are broken")
+class TestBF16Cast(unittest.TestCase):
```
Comment (chenyuxyz) [inline test/test_dtype.py:355]: `@unitted.skipUnless(is_dtype_supported(dtypes.float16))`

File: test/unit/test_disk_tensor.py [hunk 2]
```diff
@@ -31,6 +31,7 @@ def test_load_convnext(self): compare_weights_both('https://dl.fbaipublicfiles.c
   # CUDACPU architecture is sm_35 but we need at least sm_70 to run fp16 ALUs
   @unittest.skipIf(Device.DEFAULT in ["GPU", "LLVM", "CUDA"] and CI, "fp16 broken in some backends")
   @unittest.skipIf(Device.DEFAULT == "TORCH", "torch doesn't support the way we load bfloat (cast to uint32)")
+  @unittest.skipIf(Device.DEFAULT == "WEBGPU", "float16 and uint16 are broken in webgpu")
```
Comment (chenyuxyz) [inline test/unit/test_disk_tensor.py:34]: can this use `is_dtype_supported` too? can make `is_dtype_supported` take multiple dtypes then it's clear what are the types that need to support for each test

File: tinygrad/tensor.py [hunk 3]
```diff
@@ -875,9 +875,9 @@ def sparse_categorical_crossentropy(self, Y:Tensor, ignore_index=-1) -> Tensor:
   # ***** cast ops *****
 
   def cast(self, dtype:DType) -> Tensor:
-    # hack for devices that don't support bfloat16
-    if self.dtype == dtypes.bfloat16:
-      return self.bitcast(dtypes.uint16).cast(dtypes.uint32).mul(1<<16).contiguous().bitcast(dtypes.float32).cast(dtype)
+    # hack loading bf16 from disktensors for devices that don't support bfloat16
```
Comment (geohot) [inline tinygrad/tensor.py:878]: Hmm, this doesn't seem right. This is if the tensor is bfloat16, not being casted to bfloat16

File: test/unit/test_disk_tensor.py [hunk 4]
```diff
@@ -13,7 +13,7 @@ def compare_weights_both(url):
   torch_weights = get_state_dict(torch.load(fn, map_location=torch.device('cpu')), tensor_type=torch.Tensor)
   assert list(tg_weights.keys()) == list(torch_weights.keys())
   for k in tg_weights:
-    if tg_weights[k].dtype == dtypes.bfloat16: tg_weights[k] = torch_weights[k].float() # numpy doesn't support bfloat16
+    if tg_weights[k].dtype == dtypes.bfloat16: tg_weights[k] = tg_weights[k].float() # numpy doesn't support bfloat16
```
Comment (geohot) [inline test/unit/test_disk_tensor.py:16]: Huh?!?

Comment (Qazalin) [reply test/unit/test_disk_tensor.py:16]: broke after this https://github.com/tinygrad/tinygrad/pull/2698/files#diff-7884618562083b1467b1b4120d56fb3c2d5d2945d328afe4f170365932a60dedR16

File: tinygrad/tensor.py (hunk 3 above)
Comment (Qazalin) [reply tinygrad/tensor.py:875]: then we'll have to extend the hack to linearizer and ops_disk to make the cleanup work.
https://github.com/tinygrad/tinygrad/pull/2794/commits/eead41882342a15bfee2148758818403a6e506e8

File: tinygrad/runtime/ops_cpu.py [hunk 5]
```diff
@@ -27,7 +27,7 @@ def mulacc(a, b, new_shape):
 numpy_fxn_for_op: Dict[Op, Callable] = {
   BufferOps.CONST: lambda val, dtype: np.array(val, dtype=dtype.np),
   UnaryOps.EXP2: np.exp2, UnaryOps.LOG2: np.log2, UnaryOps.SIN: np.sin,
-  UnaryOps.CAST: lambda x,y: x.view(y[0].np) if y[1] else x.astype(y[0].np, copy=False),
+  UnaryOps.CAST: lambda x,y: x.view(np.uint16) if y[0] == dtypes.bfloat16 else x.view(y[0].np) if y[1] else x.astype(y[0].np, copy=False), # noqa: E501 numpy does not support bfloat16
```
Comment (geohot) [inline tinygrad/runtime/ops_cpu.py:30]: Why is there a noqa here?

And that inline if can be written better

**Comment by geohot** [comment]
I don't get what's wrong now. I don't think you need these hacks in the linearizer.

It's okay if a backend doesn't support a dtype, we have to be robust to this in a generic way.

**Comment by geohot** [comment]
I will never merge hacks that look like this, adding if statements to multiple places.

**Comment by geohot** [comment]
If a disk tensor is bfloat16, and you do .to('cpu') on it, that should just fail. CPU doesn't support BF16

**Comment by Qazalin** [comment]
I'll reopen and come back to this

**Comment by eichenroth** [comment]
That sounded promising, are you still working on a fix @Qazalin ?

**Pull request closed without merging**